compiler-user-guide/src/developing-hardware/primitives.md

```diff
@@ -0,0 +1,397 @@
+* `~LENGTH[<HOLE>]`: The vector length of the type represented by `<HOLE>`.
+* `~DEPTH[<HOLE>]`: The tree depth of the type represented by `<HOLE>`.
+  The content of `<HOLE>` must either be: `~TYP[N]`, `~TYPO`, or `~TYPEL[<HOLE>]`.
+* `~SIZE[<HOLE>]`: The number of bits needed to encode the type represented by `<HOLE>`.
+  The content of `<HOLE>` must either be: `~TYP[N]`, `~TYPO`, or `~TYPEL[<HOLE>]`.
+* `~IF <CONDITION> ~THEN <THEN> ~ELSE <ELSE> ~FI`: renders the `<ELSE>`
+  part when `<CONDITION>` evaluates to _0_, and renders the `<THEN>` in all
+  other cases.
+  Valid `<CONDITION>`s are `~LENGTH[<HOLE>]`, `~SIZE[<HOLE>]`, `~CMPLE[<HOLE1>][<HOLE2>]`, `~DEPTH[<HOLE>]`, `~VIVADO`, `~IW64`, `~ISLIT[N]`, `~ISVAR[N]`, `~ISACTIVEENABLE[N]`, `~ISSYNC[N]`, and `~AND[<HOLE1>,<HOLE2>,..]`.
+* `~VIVADO`: _1_ when Clash compiler is invoked with the `-fclash-xilinx` or
+  `-fclash-vivado` flag.
+  To be used with in an `~IF .. ~THEN .. ~ELSE .. ~FI` statement.
+* `~CMPLE[<HOLE1>][<HOLE2>]`: _1_ when `<HOLE1> <= <HOLE2>`, otherwise _0_
+* `~IW64`: _1_ when `Int`/`Word`/`Integer` types are represented with 64 bits in HDL.
+  _0_ when they're represented by 32 bits.
+* `~TOBV[<HOLE>][<TYPE>]`: create conversion code that so that the expression in `<HOLE>` is converted to a bit vector (`std_logic_vector`).
+  The `<TYPE>` hole indicates the type of the expression and must be either `~TYP[N]`, `~TYPO`, or `~TYPEL[<HOLE>]`.
+* `~FROMBV[<HOLE>][<TYPE>]`: create conversion code that so that the expression in `<HOLE>`, which has a bit vector (`std_logic_vector`) type, is converted to type indicated by `<TYPE>`.
+  The `<TYPE>` hole must be either `~TYP[N]`, `~TYPO`, or `~TYPEL[<HOLE>]`.
+* `~INCLUDENAME[N]`: the generated name of the `N`'th included component.
+* `~FILE[<HOLE>]`: The argument mentioned in `<HOLE>` is a file which must be copied to the location of the generated HDL.
+* `~GENERATE`: Verilog: create a _generate_ statement, except when already in a _generate_ context.
+* `~ENDGENERATE`: Verilog: create an _endgenerate_ statement, except when already in a _generate_ context.
+* `~ISLIT[N]`: Is the `(N+1)`'th argument to the function a literal.
+* `~ISVAR[N]`: Is the `(N+1)`'th argument to the function explicitly not a literal.
+* `~ISSCALAR[N]`: Is the `(N+1)`'th argument to the function a scalar.
+  Note that this means different things for different HDLs.
+  In (System)Verilog only `Bit` and `Bool` are considered scalar.
+  In VHDL, in addition to those two, enumeration types and integers are considered scalar.
+* `~TAG[N]`: Name of given domain.
+  Errors when called on an argument which is not a 'KnownDomain', 'Reset', or 'Clock'.
+* `~PERIOD[N]`: Clock period of given domain.
+  Errors when called on an argument which is not a 'Clock', 'Reset', 'KnownDomain' or 'KnownConf'.
+* `~ISACTIVEENABLE[N]`: Is the `(N+1)`'th argument a an Enable line __not__ set to a constant True.
+  Can be used instead of deprecated (and removed) template tag
```

Comment:
We really properly removed that template tag, didn't we? We even removed it here.